### ISSUE TITLE
EventHubTarget not sending all events before returning when NLog.LogManager.Flush() is called.

### ISSUE THREAD

**Issue by gabehunt** [JDetmar/NLog.Extensions.AzureStorage#85]
Issue:
My console application sends 100,000 log events at 1kb each.  Calling NLog.LogManager.Flush() at the end of my program returns instantly even though events are not fully sent.  This results in lost log events and many exceptions such as: Exception thrown: 'System.OperationCanceledException' in Microsoft.Azure.Amqp.dll.

Potential root cause:
Issue appears to be from this ContinueWith call:  sendTask = sendTask.ContinueWith(async p => await WriteSingleBatchAsync(batchItem, partitionKey).ConfigureAwait(false), cancellationToken);

Potential fix:
When I change that code to the following, the flush will send all events before shutting down and no exceptions are thrown:
multipleTasks.Add(WriteSingleBatchAsync(batchItem, partitionKey));

Please feel free to close my issue if it is not truly and issue and turns out to be my misunderstanding of how the async flush should work.

Thank you!
![eventHubTargetFlushIssue](https://user-images.githubusercontent.com/29798632/98950786-108d9600-24bf-11eb-8314-8eab23ae6fc2.png)

**Comment by snakefoot** [comment]
Can you share your NLog.config ?

Did you configure `overflowAction="Block"` ? Documentation says it is [Discard by default](https://github.com/JDetmar/NLog.Extensions.AzureStorage/blob/master/src/NLog.Extensions.AzureEventHub/README.md#batching-policy).

**Comment by gabehunt** [comment]
I had overflowAction set to Grow, but now that you say that it is probably better to apply back pressure on the caller than it is to let unlimited memory and threads spawn.
    <target xsi:type="AzureEventHub"            name="azureEventHub"            eventHubName="logging-prod-hub"            contentType="application/json"            queueLimit="50000" batchSize="10000" overflowAction="Grow"            connectionString="Endpoint=sb://............">      <contextProperty name="level" layout="${level}" />      <contextProperty name="exception" layout="${exception:format=shorttype}" includeEmptyValue="false" />      <layout type="JsonLayout">        <attribute name="time" layout="${longdate}" />        <attribute name="level" layout="${level:uppercase=true}" />        <attribute name="env" layout="Stage" />        <attribute name="source" layout="dotnet" />        <attribute name="correlationId" layout="" />        <attribute name="applicationName" layout=" ${assembly-name}" />        <attribute name="applicationVersion" layout="${assembly-version}" />        <attribute name="message" layout="${message}" />        <attribute name="exception" layout="${onexception:${exception:format=tostring}}" />        <attribute name="errordata" layout="${onexception:${exception:format=data}}" />        <attribute name="stacktrace" layout="${onexception:${stacktrace:format=Raw}}" />      </layout>    </target>    


    On Thursday, November 12, 2020, 09:58:17 AM CST, Rolf Kristensen <notifications@github.com> wrote:  
 
 


Can you share your NLog.config ?

Did you configure overflowAction="Block" ? Documentation says it is Discard by default.

**Comment by snakefoot** [comment]
Well tried to blast 400.000 LogEvents through EventHubTarget using an unit-test, and could not reproduce your issue.

Can you adjust this unit-test to replicate the issue you are seeing:

```c#
        [Fact]
        public void EventDataBulkBatch()
        {
            var logFactory = new LogFactory();
            var logConfig = new Config.LoggingConfiguration(logFactory);
            var eventHubService = new EventHubServiceMock();
            var eventHubTarget = new EventHubTarget(eventHubService);
            eventHubTarget.OverflowAction = Targets.Wrappers.AsyncTargetWrapperOverflowAction.Grow;
            eventHubTarget.ConnectionString = "LocalEventHub";
            eventHubTarget.PartitionKey = "${logger}";
            eventHubTarget.Layout = "${message}";
            eventHubTarget.TaskDelayMilliseconds = 200;
            eventHubTarget.BatchSize = 10000;
            eventHubTarget.QueueLimit = 50000;
            eventHubTarget.IncludeEventProperties = true;
            logConfig.AddRuleForAllLevels(eventHubTarget);
            logFactory.Configuration = logConfig;
            for (int i = 0; i < 400000; ++i)
            {
                if (i % 1000 == 0)
                    System.Threading.Thread.Sleep(1);
                logFactory.GetLogger("Test").Info("Hello {Counter}", i);
            }
            logFactory.Flush();
            Assert.Single(eventHubService.EventDataSent);
            Assert.Equal(400000, eventHubService.EventDataSent.First().Value.Count);
            var previous = 0;
            foreach (var item in eventHubService.EventDataSent.First().Value)
                if ((int)item.Properties["Counter"] > previous)
                    previous = (int)item.Properties["Counter"];
        }
```

P.S. Have to adjust the `EventHubServiceMock` a little to handle the above test:

```c#
    class EventHubServiceMock : IEventHubService
    {
        public Dictionary<string, List<EventData>> EventDataSent { get; } = new Dictionary<string, List<EventData>>();

        public Task SendAsync(IList<EventData> eventDataList, string partitionKey)
        {
            if (string.IsNullOrEmpty(ConnectionString))
                throw new InvalidOperationException("EventHubService not connected");

            lock (EventDataSent)
            {
                if (EventDataSent.TryGetValue(partitionKey, out var existingBatch))
                    existingBatch.AddRange(eventDataList);
                else
                    EventDataSent[partitionKey] = new List<EventData>(eventDataList);
            }
            return Task.Delay(10);
        }
```

**Comment by gabehunt** [comment]
Interesting.  I was able to send 200,000 through the unit test no problem.  However, the actual console app sends 100,000 to a real Azure Premium Event Hub with 4 messaging units and 32 partitions and whether I set Block or Grow in nlog.config the console app closes almost immediately after calling NLog.LogManager.Flush(120000);
*Note: The release version works great as-is in any long running app (Hosted Web App, Hosted Service, etc.).  After the endpoint of an api returns data to the caller that logger stays running for almost 1 minute in the background and we get all the events.
    On Thursday, November 12, 2020, 02:50:50 PM CST, Rolf Kristensen <notifications@github.com> wrote:  
 
 


Well tried to blast 400.000 LogEvents through EventHubTarget using an unit-test, and could not reproduce your issue.

Can you adjust this unit-test to replicate the issue you are seeing:
        [Fact]
        public void EventDataBulkBatch()
        {
            var logFactory = new LogFactory();
            var logConfig = new Config.LoggingConfiguration(logFactory);
            var eventHubService = new EventHubServiceMock();
            var eventHubTarget = new EventHubTarget(eventHubService);
            eventHubTarget.OverflowAction = Targets.Wrappers.AsyncTargetWrapperOverflowAction.Grow;
            eventHubTarget.ConnectionString = "LocalEventHub";
            eventHubTarget.PartitionKey = "${logger}";
            eventHubTarget.Layout = "${message}";
            eventHubTarget.TaskDelayMilliseconds = 200;
            eventHubTarget.BatchSize = 10000;
            eventHubTarget.QueueLimit = 50000;
            eventHubTarget.IncludeEventProperties = true;
            logConfig.AddRuleForAllLevels(eventHubTarget);
            logFactory.Configuration = logConfig;
            for (int i = 0; i < 400000; ++i)
            {
                if (i % 1000 == 0)
                    System.Threading.Thread.Sleep(1);
                logFactory.GetLogger("Test").Info("Hello {Counter}", i);
            }
            logFactory.Flush();
            Assert.Single(eventHubService.EventDataSent);
            Assert.Equal(400000, eventHubService.EventDataSent.First().Value.Count);
            var previous = 0;
            foreach (var item in eventHubService.EventDataSent.First().Value)
                if ((int)item.Properties["Counter"] > previous)
                    previous = (int)item.Properties["Counter"];
        }
P.S. Have to adjust the EventHubServiceMock a little to handle the above test:
    class EventHubServiceMock : IEventHubService
    {
        public Dictionary<string, List<EventData>> EventDataSent { get; } = new Dictionary<string, List<EventData>>();

        public Task SendAsync(IList<EventData> eventDataList, string partitionKey)
        {
            if (string.IsNullOrEmpty(ConnectionString))
                throw new InvalidOperationException("EventHubService not connected");

            lock (EventDataSent)
            {
                if (EventDataSent.TryGetValue(partitionKey, out var existingBatch))
                    existingBatch.AddRange(eventDataList);
                else
                    EventDataSent[partitionKey] = new List<EventData>(eventDataList);
            }
            return Task.Delay(10);
        }

**Comment by snakefoot** [comment]
I was able to find a bug, while trying to reproduce your issue, but that bug is about using `overflow="Block"` and flushing when the queue is completely full:

https://github.com/NLog/NLog/pull/4172

But your initial case is `overflow="Block"`, so you are probably experiencing another issue. Could you try and enable the NLog-InternalLogger when doing the flush like this:

```c#
NLog.Common.InternalLogger.LogLevel = NLog.LogLevel.Debug;
NLog.Common.InternalLogger.LogFile = "your-favorite-location.txt";
NLog.LogManager.Flush(120000);
```

You should atleast see two messages:

```c#
InternalLogger.Debug("{0} Flushing {1}", Name, _requestQueue.IsEmpty ? "empty queue" : "pending queue items");
InternalLogger.Debug("{0} Flush Completed", Name);
```

**Comment by snakefoot** [comment]
Was able to reproduce the issue when using BatchSize > 100. Updated PullRequest with fix:  #83

**Comment by gabehunt** [comment]
Thank you @snakefoot for such a quick and diligent response on this!!

**Comment by snakefoot** [comment]
@gabehunt You are welcome. Is a little embarrasing bug, but atleast I learned something by reading this:

https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-await-over-continuewith

That I'm not clever enough to use ContinueWith, and should just relax and eat the cake,

**Comment by JDetmar** [comment]
thank you @snakefoot and @gabehunt 
the update has been pushed to nuget and should be available soon.

**Comment by snakefoot** [comment]
Yes ver. 2.6.0 is now available: https://www.nuget.org/packages/NLog.Extensions.AzureEventHub/

**Comment by snakefoot** [comment]
The embarassing bug with `overflow="Block"` and explicit flush have now been fixed with NLog 4.7.6:

https://www.nuget.org/packages/NLog/4.7.6
